File: backend/services/books/handlers.go
package books

import (
	"fmt"
	"math"
	"net/http"

	"github.com/ggicci/httpin"
	"github.com/mrsafalpiya/library-management/server"
	"github.com/mrsafalpiya/library-management/utils"
)

func handleListBooks(srvCfg *server.Config) http.HandlerFunc {
	paginationMeta := struct {
		CurrentPage      int `json:"current_page"`
		PageSize         int `json:"page_size"`
		FirstPage        int `json:"first_page"`
		LastPage         int `json:"last_page"`
		TotalRecords     int `json:"total_records"`
		PageRecordsCount int `json:"page_records_count"`
	}{}

	return func(w http.ResponseWriter, r *http.Request) {
		input := r.Context().Value(httpin.Input).(*ListBooksQuery)

		paginationMeta.PageRecordsCount = 0

		// Set sort

		sortQuery1, sortQuery2, err := utils.GetSortQuery(input.Sort, "code", "title", "author", "publisher")
		if err == utils.ErrInvalidSortQuery {
			utils.ResponseBadRequestErr(w, "invalid sort query")
			return
		}

		query := fmt.Sprintf(`
			SELECT *
			FROM books
			WHERE to_tsvector('english', title) @@ to_tsquery('english', $3) or $3 = ''
			ORDER BY %s %s
			LIMIT $1
			OFFSET $2;`, sortQuery1, sortQuery2)
		args := []any{
			input.Size,
			(input.Page - 1) * input.Size,
			input.Search,
		}

		rows, err := srvCfg.DbConn.Query(query, args...)
		if err != nil {
			utils.ResponseBadRequestErr(w, err.Error())
			return
		}

		type Book struct {
			ID        int64  `json:"id"`
			Code      string `json:"code"`
			Title     string `json:"title"`
			Author    string `json:"author"`
			Publisher string `json:"publisher"`
		}

		books := []Book{}
		for rows.Next() {
			book := Book{}

			err := rows.Scan(&book.ID, &book.Code, &book.Title, &book.Author, &book.Publisher)
			if err != nil {
				utils.ResponseBadRequestErr(w, err.Error())
				return
			}

			books = append(books, book)
			paginationMeta.PageRecordsCount++
		}

		// Query for pagination metadata

		query = `
			SELECT COUNT(*)
			FROM books
			WHERE to_tsvector('english', title) @@ to_tsquery('english', $1) or $1 = ''`
		row := srvCfg.DbConn.QueryRow(query, input.Search)
		err = row.Scan(&paginationMeta.TotalRecords)
		if err != nil {
			utils.ResponseBadRequestErr(w, err.Error())
			return
		}

		// Calculate pagination metadata

		paginationMeta.CurrentPage = int(input.Page)
		paginationMeta.PageSize = int(input.Size)
		paginationMeta.FirstPage = 1
		paginationMeta.LastPage = int(math.Ceil(float64(paginationMeta.TotalRecords) / float64(paginationMeta.PageSize)))

		utils.ResponseOKData(w, utils.Envelope{
			"metadata": paginationMeta,
			"books":    books,
		})
	}
}

File: backend/src/services/books/handlers.ts
import { ServerConfig } from "app";
import { Response, RequestHandler, Request } from "express";
import { DatabaseError } from "pg";
import {
  responseBadRequest,
  responseOK,
  responseServerError,
} from "server/json";
import { PaginationMetadata } from "server/pagination";
import { getSortQuery } from "server/utils";
import {
  getCopy,
  IListBooksResult,
  issueBookBorrows,
  issueBookTransactions,
  listBooks,
  listBooksCount,
  renewBookBorrows,
  renewBookTransactions,
  returnBookBorrows,
  returnBookTransactions,
} from "./queries";

export function handleListBooks(serverCfg: ServerConfig): RequestHandler {
  return async function (req: Request, res: Response) {
    const querySize = parseInt(req.query.size as string) || 10;
    const queryPage = parseInt(req.query.page as string) || 1;
    const querySort = (req.query.sort as string) || "title-asc";
    const querySearch = (req.query.search as string) || "";

    // Check if the search query is valid, also parse it for query
    let orderParam, orderSort;
    try {
      [orderParam, orderSort] = getSortQuery(querySort, [
        "title",
        "author",
        "publisher",
      ]);
    } catch (e) {
      const err = e as Error;
      responseBadRequest(res, err.message);
      return;
    }

    // Query books

    let books: IListBooksResult[];

    try {
      books = await listBooks.run(
        {
          searchParam: querySearch,
          orderParam: orderParam,
          orderSort: orderSort,
          limit: querySize,
          offset: (queryPage - 1) * querySize,
        },
        serverCfg.dbConn
      );
    } catch (e) {
      responseServerError(res, e);
      return;
    }

    // Query for pagination metadata

    let paginationMetadata: PaginationMetadata = {
      current_page: queryPage,
      first_page: 1,
      last_page: 0,
      page_records_count: 0,
      page_size: querySize,
      total_records: 0,
    };

    try {
      const queryResult = await listBooksCount.run(
        { searchParam: querySearch },
        serverCfg.dbConn
      );
      if (queryResult[0].count == null) {
        responseServerError(
          res,
          "somehow pagination count query response is null"
        );
        return;
      }
      paginationMetadata.page_records_count = books.length;
      paginationMetadata.total_records = parseInt(queryResult[0].count);
      paginationMetadata.last_page = Math.ceil(
        paginationMetadata.total_records / querySize
      );
    } catch (e) {
      responseServerError(res, e);
      return;
    }

    responseOK(res, {
      metadata: paginationMetadata,
      books: books,
    });
  };
}

export function handleGetCopy(serverCfg: ServerConfig): RequestHandler {
  return async function (req: Request, res: Response) {
    const registerID = req.params.registerID;

    try {
      const queryResponse = await getCopy.run({ registerID }, serverCfg.dbConn);

      if (queryResponse.length == 0) {
        responseBadRequest(res, "given registration ID does not exist");
        return;
      }

      responseOK(res, { book: queryResponse[0] });
      return;
    } catch (e) {
      responseServerError(res, e);
      return;
    }
  };
}

export function handleIssueCopy(serverCfg: ServerConfig): RequestHandler {
  return async function (req: Request, res: Response) {
    try {
      await serverCfg.dbConn.query("BEGIN");

      await issueBookBorrows.run(
        {
          copyID: req.body.copy_id,
          studentID: req.body.student_id,
          durationDays: req.body.issue_duration_days,
        },
        serverCfg.dbConn
      );
      await issueBookTransactions.run(
        {
          copyID: req.body.copy_id,
          studentID: req.body.student_id,
        },
        serverCfg.dbConn
      );

      await serverCfg.dbConn.query("COMMIT");

      responseOK(res, { response: "ok" });
      return;
    } catch (e) {
      await serverCfg.dbConn.query("ROLLBACK;");

      const error = e as DatabaseError;

      switch (error.code) {
        case "23505":
          responseBadRequest(res, "the student has already borrowed this book");
          break;
        default:
          responseServerError(res, e);
      }
      return;
    }
  };
}

export function handleReturnCopy(serverCfg: ServerConfig): RequestHandler {
  return async function (req: Request, res: Response) {
    try {
      await serverCfg.dbConn.query("BEGIN");

      await returnBookBorrows.run(
        { copyID: req.body.copy_id },
        serverCfg.dbConn
      );
      await returnBookTransactions.run(
        { copyID: req.body.copy_id, studentID: req.body.student_id },
        serverCfg.dbConn
      );

      await serverCfg.dbConn.query("COMMIT;");

      responseOK(res, { response: "ok" });
      return;
    } catch (e) {
      await serverCfg.dbConn.query("ROLLBACK;");

      const error = e as DatabaseError;

      switch (error.code) {
        default:
          responseServerError(res, e);
      }

      return;
    }
  };
}

export function handleRenew(serverCfg: ServerConfig): RequestHandler {
  return async function (req: Request, res: Response) {
    try {
      await serverCfg.dbConn.query("BEGIN");

      await renewBookBorrows.run(
        {
          durationDays: req.body.issue_duration_days,
          copyID: req.body.copy_id,
        },
        serverCfg.dbConn
      );
      await renewBookTransactions.run(
        { copyID: req.body.copy_id, studentID: req.body.student_id },
        serverCfg.dbConn
      );

      await serverCfg.dbConn.query("COMMIT;");

      responseOK(res, { response: "ok" });
      return;
    } catch (e) {
      await serverCfg.dbConn.query("ROLLBACK;");

      const error = e as DatabaseError;

      switch (error.code) {
        default:
          responseServerError(res, e);
      }

      return;
    }
  };
}
